clc/modules/object-storage-common/src/main/java/com/eucalyptus/objectstorage/util/OSGUtil.java

```diff
@@ -333,4 +349,144 @@ public static String getBucketFromHostHeader(MappingHttpRequest httpRequest) thr
     }
     return hostBucket;
   }
+  
+  public static CorsMatchResult matchCorsRules (List<CorsRule> corsRules, String requestOrigin, 
```

Comment:
The conditional processing seemed a little difficult to follow. Branching based on the boolean variable found could be error prone. How about a simplified version using if-else blocks and a little bit of guava for handling loops
`public static CorsMatchResult matchCorsRules(List<CorsRule> corsRules, String requestOrigin, String requestMethod, List<String> requestHeaders) {
    CorsMatchResult corsMatchResult = new CorsMatchResult();
    CorsRule corsRuleMatch = null;

    // Predicate for matching origin
    Predicate<String> originMatch = new Predicate<String>() {

      @Override
      public boolean apply(String allowedOrigin) {
        String allowedOriginRegex = "\\Q" + allowedOrigin.replace("*", "\\E.*?\\Q") + "\\E";
        return Pattern.matches(allowedOriginRegex, requestOrigin);
      }
    };

    // Predicate for matching method
    Predicate<String> methodMatch = new Predicate<String>() {

      @Override
      public boolean apply(String allowedMethod) {
        return requestMethod.equals(allowedMethod);
      }
    };

    // Predicate for generating a pattern from an allowed header
    Function<String, Pattern> generatePattern = new Function<String, Pattern>() {

      @Override
      public Pattern apply(String allowedHeader) {
        String allowedHeaderRegex = "\\Q" + allowedHeader.replace("*", "\\E.*?\\Q") + "\\E";
        return Pattern.compile(allowedHeaderRegex, Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE);
      }
    };

    for (CorsRule corsRule : corsRules) {

      if (corsRule == null) {
        continue;
      }

      String matchingOrigin = null;

      // Does the origin match any origin's regular expression in the rule?
      if (corsRule.getAllowedOrigins() != null && !corsRule.getAllowedOrigins().isEmpty()
          && (matchingOrigin = Iterables.tryFind(corsRule.getAllowedOrigins(), originMatch).orNull()) != null) {

        // Does the HTTP verb match any verb in the rule?
        if (corsRule.getAllowedMethods() != null && !corsRule.getAllowedMethods().isEmpty()
            && Iterables.any(corsRule.getAllowedMethods(), methodMatch)) {

          // If there are no Access-Control-Request-Headers, or if there are
          // no AllowedHeaders in the CORS rule, then skip this check.
          if (requestHeaders == null || requestHeaders.isEmpty() || corsRule.getAllowedHeaders() == null || corsRule.getAllowedHeaders().isEmpty()) {

            corsMatchResult.setCorsRuleMatch(corsRuleMatch);
            corsMatchResult.setAnyOrigin(matchingOrigin.equals("*"));
            break; // We have matched the current CORS rule. Stop looking through them.

          } else {

            List<String> allowedHeaders = corsRule.getAllowedHeaders();
            List<Pattern> allowedHeaderPatternList = new ArrayList<Pattern>(corsRule.getAllowedHeaders().size());

            // Predicate for matching request header with allowed headers
            Predicate<String> headerMatch = new Predicate<String>() {

              @Override
              public boolean apply(String requestHeader) {
                for (int idx = 0; idx < allowedHeaders.size(); idx++) {
                  Pattern allowedHeaderPattern;
                  if ((allowedHeaderPattern = allowedHeaderPatternList.get(idx)) == null) {
                    allowedHeaderPattern = generatePattern.apply(allowedHeaders.get(idx));
                    allowedHeaderPatternList.add(allowedHeaderPattern);
                  }
                  Matcher matcher = allowedHeaderPattern.matcher(requestHeader);
                  if (matcher.matches()) {
                    return true; // stop looking through the allowed headers for this request header
                  } else {
                    continue; // try matching request header with allowed header until a match is found
                  }
                }
                return false; // No allowed header matches this request header, so this rule fails to match
              }
            };

            // Does every request header in the comma-delimited list in Access-Control-Request-Headers have a matching entry in the allowed headers in
            // the rule?
            // Headers are matched case-insensitively.s
            if (Iterables.all(requestHeaders, headerMatch)) {

              corsMatchResult.setCorsRuleMatch(corsRuleMatch);
              corsMatchResult.setAnyOrigin(matchingOrigin.equals("*"));
              break; // We have matched the current CORS rule. Stop looking through them.

            } else {
              continue; // go to the next CORS rule
            }
          }
        } else {
          continue; // go to the next CORS rule
        }
      } else {
        continue; // go to the next CORS rule
      }
    }

    return corsMatchResult;
  }`